File: shared/elevenlabs.py
"""ElevenLabs TTS (mirror of server/elevenlabs.go)."""
from __future__ import annotations
import httpx
from .config import ELEVEN_LABS_API_KEY, ELEVEN_LABS_VOICE, ELEVEN_LABS_MODEL


class TTSError(Exception):
    pass


def tts(text: str) -> bytes:
    if not ELEVEN_LABS_API_KEY:
        raise TTSError("TTS_API_KEY (ElevenLabs) not set")
    voice = ELEVEN_LABS_VOICE or "JBFqnCBsd6RMkjVDRZzb"
    payload = {"text": text, "model_id": ELEVEN_LABS_MODEL,
               "voice_settings": {"stability": 0.5, "similarity_boost": 0.75}}
    headers = {"xi-api-key": ELEVEN_LABS_API_KEY, "Content-Type": "application/json",
               "Accept": "audio/mpeg"}
    try:
        r = httpx.post(f"https://api.elevenlabs.io/v1/text-to-speech/{voice}",
                       json=payload, headers=headers, timeout=120.0)
    except httpx.HTTPError as e:
        raise TTSError(str(e))
    if r.status_code != 200:
        raise TTSError(f"elevenlabs {r.status_code}: {r.text[:300]}")
    return r.content

File: server/elevenlabs.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// tts calls ElevenLabs text-to-speech and returns MP3 bytes.
func (a *App) tts(text string) ([]byte, error) {
	if a.cfg.ElevenLabsKey == "" {
		return nil, fmt.Errorf("TTS_API_KEY (ElevenLabs) not set")
	}
	voice := a.cfg.ElevenLabsVoice
	if voice == "" {
		voice = "JBFqnCBsd6RMkjVDRZzb"
	}
	body, _ := json.Marshal(map[string]any{
		"text":     text,
		"model_id": a.cfg.ElevenLabsModel,
		"voice_settings": map[string]any{
			"stability":        0.5,
			"similarity_boost": 0.75,
		},
	})
	url := "https://api.elevenlabs.io/v1/text-to-speech/" + voice
	req, _ := http.NewRequest("POST", url, bytes.NewReader(body))
	req.Header.Set("xi-api-key", a.cfg.ElevenLabsKey)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "audio/mpeg")
	resp, err := (&http.Client{Timeout: 120 * time.Second}).Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	data, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("elevenlabs %d: %s", resp.StatusCode, truncate(string(data), 300))
	}
	return data, nil
}
